src/darwin/Framework/CHIP/MTRDevice.mm

```diff
@@ -235,7 +267,38 @@ - (void)_handleResubscriptionNeeded
 
 - (void)_handleSubscriptionReset
 {
-    // TODO: logic to reattempt subscription with exponential back off
+    os_unfair_lock_lock(&self->_lock);
+    // if there is no delegate then also do not retry
+    id<MTRDeviceDelegate> delegate = _weakDelegate.strongObject;
+    if (!delegate) {
+        MTR_LOG_INFO("%@ no delegate - do not reattempt subscription", self);
+        return;
+    }
+
+    // don't schedule multiple retries
+    if (self.reattemptingSubscription) {
+        MTR_LOG_INFO("%@ already reattempting subscription", self);
+        return;
+    }
+
+    self.reattemptingSubscription = YES;
+
+    if (_lastSubscriptionAttemptWait < MTRDEVICE_SUBSCRIPTION_ATTEMPT_MIN_WAIT) {
+        _lastSubscriptionAttemptWait = MTRDEVICE_SUBSCRIPTION_ATTEMPT_MIN_WAIT;
+    } else {
+        _lastSubscriptionAttemptWait *= 2;
+    }
+
+    MTR_LOG_INFO("%@ scheduling to reattempt subscription in %u seconds", self, _lastSubscriptionAttemptWait);
+    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(_lastSubscriptionAttemptWait * NSEC_PER_SEC)), self.queue, ^{
```

Comment:
```suggestion
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, static_cast<int64_t>(_lastSubscriptionAttemptWait) * NSEC_PER_SEC), self.queue, ^{
```
seems like the right thing... Otherwise the multiplication can overflow before we ever cast to 64-bit, depending on what the type of NSEC_PER_SEC is.